// angular4-material-table/src/table-element.js
import cloneDeep from 'lodash.clonedeep';
var TableElement = /** @class */ (function () {
    function TableElement(init) {
        Object.assign(this, init);
    }
    TableElement.prototype.delete = function () {
        this.source.delete(this.id);
    };
    TableElement.prototype.confirmEditCreate = function () {
        this.originalData = undefined;
        if (this.id == -1)
            return this.source.confirmCreate(this);
        else
            return this.source.confirmEdit(this);
    };
    TableElement.prototype.startEdit = function () {
        this.originalData = cloneDeep(this.currentData);
        this.editing = true;
        this.validator.enable();
    };
    TableElement.prototype.cancelOrDelete = function () {
        if (this.id == -1 || !this.editing)
            this.delete();
        else {
            this.currentData = this.originalData;
            this.editing = false;
            this.validator.disable();
        }
    };
    return TableElement;
}());
export { TableElement };
